src/script/sigcommon.cpp

```diff
@@ -53,68 +53,233 @@ uint256 GetOutputsHash(const CTransaction &txTo)
+                itBegin = it;
+            }
+        }
+        if (itBegin != scriptCode.end())
+            s.write((char *)&itBegin[0], it - itBegin);
+    }
+
+    /** Serialize an input of txTo */
+    template <typename S>
+    void SerializeInput(S &s, unsigned int nInput) const
+    {
+        // In case of SIGHASH_ANYONECANPAY, only the input being signed is serialized
+        if (fAnyoneCanPay)
+            nInput = nIn;
+        // Serialize the prevout
+        ::Serialize(s, txTo.vin[nInput].prevout);
+        // Serialize the script
+        if (nInput != nIn)
+            // Blank out other inputs' signatures
+            ::Serialize(s, CScriptBase());
+        else
+            SerializeScriptCode(s);
+        // Serialize the nSequence
+        if (nInput != nIn && (fHashSingle || fHashNone))
+            // let the others update at will
+            ::Serialize(s, (int)0);
+        else
+            ::Serialize(s, txTo.vin[nInput].nSequence);
+    }
+
+    /** Serialize an output of txTo */
+    template <typename S>
+    void SerializeOutput(S &s, unsigned int nOutput) const
+    {
+        if (fHashSingle && nOutput != nIn)
+            // Do not lock-in the txout payee at other indices as txin
+            ::Serialize(s, CTxOut());
+        else
+            ::Serialize(s, txTo.vout[nOutput]);
+    }
+
+    /** Serialize txTo */
+    template <typename S>
+    void Serialize(S &s) const
+    {
+        // Serialize nVersion
+        ::Serialize(s, txTo.nVersion);
+        // Serialize vin
+        unsigned int nInputs = fAnyoneCanPay ? 1 : txTo.vin.size();
+        ::WriteCompactSize(s, nInputs);
+        for (unsigned int nInput = 0; nInput < nInputs; nInput++)
+            SerializeInput(s, nInput);
+        // Serialize vout
+        unsigned int nOutputs = fHashNone ? 0 : (fHashSingle ? nIn + 1 : txTo.vout.size());
+        ::WriteCompactSize(s, nOutputs);
+        for (unsigned int nOutput = 0; nOutput < nOutputs; nOutput++)
+            SerializeOutput(s, nOutput);
+        // Serialize nLockTime
+        ::Serialize(s, txTo.nLockTime);
+    }
+};
+
 } // end anon namespace
 
-uint256 SignatureHashBitcoinCash(const CScript &scriptCode,
+// WARNING: Never use this to signal errors in a signature hash function. This is here solely for legacy reasons!
+const uint256 SIGNATURE_HASH_ERROR(uint256S("0000000000000000000000000000000000000000000000000000000000000001"));
+
+uint256 SignatureHashLegacy(const CScript &scriptCode,
     const CTransaction &txTo,
     unsigned int nIn,
     uint32_t nHashType,
     const CAmount &amount,
     size_t *nHashedOut)
 {
-    static const uint256 one(uint256S("0000000000000000000000000000000000000000000000000000000000000001"));
-
-    if (nHashType & SIGHASH_FORKID)
+    if (nIn >= txTo.vin.size())
     {
-        uint256 hashPrevouts;
-        uint256 hashSequence;
-        uint256 hashOutputs;
+        //  nIn out of range
+        // Important notice:
```

Comment:
Saw you moved the code under `sigcommon.cpp`, still I'd put this comment line in upper case.

Reply:
Alright, fixed.